pkg/secrets/keyvault/client.go

```diff
@@ -254,50 +251,67 @@ func (k *KeyvaultSecretClient) Upsert(ctx context.Context, key types.NamespacedN
 			}
 		}
 		// If flatten has not been declared, convert the map into a json string for perisstence
-	} else {
-		jsonData, err := json.Marshal(data)
-		stringSecret := string(jsonData)
-
-		// Initialize secret parameters
-		secretParams := keyvaults.SecretSetParameters{
-			Value:            &stringSecret,
-			SecretAttributes: &secretAttributes,
-		}
-
-		/*if _, err := k.KeyVaultClient.GetSecret(ctx, vaultBaseURL, secretBaseName, secretVersion); err == nil {
-			// If secret exists we delete it and recreate it again
-			_, err = k.KeyVaultClient.DeleteSecret(ctx, vaultBaseURL, secretBaseName)
-			if err != nil {
-				return fmt.Errorf("Upsert failed: Trying to delete existing secret failed with %v", err)
-			}
-		}*/
-
-		_, err = k.KeyVaultClient.SetSecret(ctx, vaultBaseURL, secretBaseName, secretParams)
+	}
+	jsonData, err := json.Marshal(data)
+	stringSecret := string(jsonData)
 
-		return err
+	// Initialize secret parameters
+	secretParams := keyvaults.SecretSetParameters{
+		Value:            &stringSecret,
+		SecretAttributes: &secretAttributes,
 	}
 
-	return nil
+	_, err = k.KeyVaultClient.SetSecret(ctx, vaultBaseURL, secretBaseName, secretParams)
+
+	return err
 }
 
 // Delete deletes a key in KeyVault
-func (k *KeyvaultSecretClient) Delete(ctx context.Context, key types.NamespacedName) error {
+func (k *KeyvaultSecretClient) Delete(ctx context.Context, key types.NamespacedName, opts ...secrets.SecretOption) error {
 	vaultBaseURL := getVaultsURL(ctx, k.KeyVaultName)
+
+	options := &secrets.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
 	var secretName string
 	if len(key.Namespace) != 0 {
 		secretName = key.Namespace + "-" + key.Name
 	} else {
 		secretName = key.Name
 	}
-	_, err := k.KeyVaultClient.DeleteSecret(ctx, vaultBaseURL, secretName)
 
+	secretVersion := ""
+	data := map[string]string{}
+	result, err := k.KeyVaultClient.GetSecret(ctx, vaultBaseURL, secretName, secretVersion)
+	if err != nil {
+		return fmt.Errorf("secret does not exist" + err.Error())
+	}
+
+	stringSecret := *result.Value
+	// ignore the error here as it generally means the value is not json and therefore cannot be traversed for additional deletions
+	_ = json.Unmarshal([]byte(stringSecret), &data)
+
+	_, err = k.KeyVaultClient.DeleteSecret(ctx, vaultBaseURL, secretName)
 	if err != nil {
 		azerr := errhelp.NewAzureErrorAzureError(err)
 		if azerr.Type != errhelp.SecretNotFound { // If not found still need to purge
 			return err
 		}
 	}
 
+	if options.Flatten {
+		for formatName := range data {
+			sName := secretName + "-" + formatName
+			_, err := k.KeyVaultClient.DeleteSecret(ctx, vaultBaseURL, sName)
+			if err != nil {
+				return err
```

Comment:
should this instead try to delete as many secrets as it can instead of aborting on first failure?